scipy/linalg/_solvers.py

```diff
@@ -302,71 +313,149 @@ def solve_discrete_are(a, b, q, r):
+    as described in [1]_, .. math:`H - \lambda J` given by the block matrices
+        
+    
+           [  A   0   B ]             [ I   0   B ] 
+           [ -Q   I   0 ] - \lambda * [ 0  A^T  0 ]
+           [  0   0   R ]             [ 0 -B^T  0 ]
+        
+    and using a QZ decomposition method.
+    
+    In this algorithm, the fail conditions are linked to the symmetrycity 
+    of the product .. math:`U_2 U_1^{-1}` and condition number of 
+    .. math:`U_1`. Here, .. math:`U` is the 2m-by-m matrix that holds the 
+    eigenvectors spanning the stable subspace with 2m rows and partitioned 
+    into two m-row matrices. See [1]_ and [2]_ for more details.
+    
 
     .. versionadded:: 0.11.0
 
+    References
+    ----------
+    .. [1]  P. van Dooren , "A Generalized Eigenvalue Approach For Solving 
+       Riccati Equations.", SIAM Journal on Scientific and Statistical 
+       Computing, Vol.2(2), DOI: 10.1137/0902010
+       
+    .. [2] Alan J Laub, "A Schur Method for Solving Algebraic Riccati 
+       Equations.", Massachusetts Institute of Technology. Laboratory for 
+       Information and Decision Systems. LIDS-R ; 859. Available online :
+       http://hdl.handle.net/1721.1/1301
+       
     """
-
-    try:
-        g = inv(r)
-    except LinAlgError:
-        raise ValueError('Matrix R in the algebraic Riccati equation solver '
-                         'is ill-conditioned')
-
-    g = np.dot(np.dot(b, g), b.conj().transpose())
-
-    try:
-        ait = inv(a).conj().transpose()  # ait is "A inverse transpose"
-    except LinAlgError:
-        raise ValueError('Matrix A in the algebraic Riccati equation solver '
-                         'is ill-conditioned')
-
-    z11 = a+np.dot(np.dot(g, ait), q)
-    z12 = -1.0*np.dot(g, ait)
-    z21 = -1.0*np.dot(ait, q)
-    z22 = ait
-
-    z = np.vstack((np.hstack((z11, z12)), np.hstack((z21, z22))))
-
-    # Note: we need to sort the upper left of s to lie within the unit circle,
-    #       while the lower right is outside (Laub, p. 7)
-    s, u, _ = schur(z, sort='iuc')
-
-    (m, n) = u.shape
-
-    u11 = u[0:m//2, 0:n//2]
-    u21 = u[m//2:m, 0:n//2]
-    u11i = inv(u11)
-
-    return np.dot(u21, u11i)
+    
+    a = _asarray_validated(a)
+    b = _asarray_validated(b)
+    q = _asarray_validated(q)
+    r = _asarray_validated(r)
+    
+    # Get the correct data types otherwise Numpy complains about pushing 
+    # complex numbers into real arrays.
+    r_or_c = complex if np.iscomplexobj(b) else float
+
+    for ind, x in enumerate((a,q,r)):
+        if np.iscomplexobj(x):
+            r_or_c = complex
+            
+        if not np.equal(*x.shape):
+            raise ValueError("Matrix {} should be square.".format("aqr"[ind]))
+    
+    # Shape consistency checks
+    m, n = b.shape
+
+    if m != a.shape[0]:
+        raise ValueError("Matrix a and b should have the same number of rows.")
+        
```

Comment:
It's not a show-stopper, but I'd like related groups of if statements to not have any spaces between them (and maybe be converted to if/elif's) so that they cluster together visually. But this might just be a personal preference so I won't press the point.